### PR TITLE
Gridworld

### PR DESCRIPTION
**Pull request by anna-ssi**
My PDM acts weird again, so couldn't pass the test when running `pdm run ./scripts/run_test.sh` ((

### CODE REVIEW

File: red_blue_world/Patch.py [hunk 1]
```diff
@@ -21,5 +21,5 @@ def load(self) -> None:
     def on_enter(self, last_state: np.ndarray) -> None:
         ...
 
-    def step(self, action: Action) -> Tuple[State, Reward, Direction]:
+    def step(self, action: Action): # -> Tuple[State, Reward, Direction]:
```
Comment (andnp) [inline red_blue_world/Patch.py]: I _think_ the latest version of mypy should be okay with this type stub. We should be okay to add the return type definition back.

File: red_blue_world/patches/gw.py [hunk 2]
```diff
@@ -1,9 +1,80 @@
...
 from red_blue_world.Patch import Patch
-from typing import Tuple
+from typing import Tuple, NamedTuple
+
+RIGHT = 0
+DOWN = 1
+LEFT = 2
+UP = 3
+
+BEAN = 0
```
Comment (andnp) [inline red_blue_world/patches/gw.py]: We might consider a different theme for our objects, just to make sure we aren't stepping on JellyBean world _too_ much. 

What's another cute theme? I kinda like the idea of picking flowers and avoiding weeds. It also makes "growing" new flowers in old patches make sense. Flowers can also emit an aroma to capture the "scent" features that JBW uses.

File: red_blue_world/patches/gw.py [hunk 3]
```diff
@@ -1,9 +1,80 @@
...
+        choosen_coords = np.random.choice(self._cell_num, object_num)
+        labels = np.random.choice([BEAN, ONION], size=object_num)
+        return choosen_coords, labels
+    
+    def _get_config(self, choosen_coords: np.ndarray, labels: np.ndarray):
+        """
+            Returns a list of PatchConfig objects where each corresponds to 
+            the object label and coordinates.
+        """
+        config = np.zeros(len(choosen_coords), dtype=PatchConfig)
```
Comment (andnp) [inline red_blue_world/patches/gw.py]: This looks good for now. Later, we might want to do a performance pass. A datatype that allows O(1) lookup for a coordinate will be much faster than O(n) lookup.

For instance, tuples of ints are hashable, so:
```python
config = {}
config[coords] = PatchConfig(...)
```
would make existence checking constant time:
```
def take_action(...):
  if (x, y) in self._config:
    # handle collision
```

File: red_blue_world/patches/gw.py [hunk 4]
```diff
@@ -1,9 +1,80 @@
...
+        for i, coord in enumerate(choosen_coords):
+            x = coord // self._size
+            y = coord % self._size
```
Comment (andnp) [inline red_blue_world/patches/gw.py]: Just out of curiosity, what's the difference in bounds-checking here? It looks like x-axis coords are clipped and y-axis coords wrap? Is that the intended behavior?

**Review by andnp — COMMENTED** [review]
Looks good! Don't worry about tests passing right now, I'll fix the test scripts in another PR.

File: red_blue_world/Patch.py (hunk 1 above)
Comment (anna-ssi) [reply red_blue_world/Patch.py]: I updated my mypy version to the latest one, but I still get an error when running `pdm run ./scripts/run_tests.sh`. Though I will remove the comment as at least now I know where the error is coming from :D

File: red_blue_world/patches/gw.py (hunk 2 above)
Comment (anna-ssi) [reply red_blue_world/patches/gw.py]: I like that idea, I will change it to flowers and weeds.

File: red_blue_world/patches/gw.py (hunk 3 above)
Comment (anna-ssi) [reply red_blue_world/patches/gw.py]: Yes, great idea! I will change the array to a dictionary now, and will change the `take_action` function when we come up with the logistics of the whole system on Thursday.

File: red_blue_world/patches/gw.py (hunk 4 above)
Comment (anna-ssi) [reply red_blue_world/patches/gw.py]: Not particularly. I think I just iterate row wise in my mind so I wrote it in this order. But I do think it is not intuitive so will swap them

**Comment by anna-ssi** [comment]
I will open a new pull request with the changes!

File: red_blue_world/patches/gw.py (hunk 4 above)
Comment (andnp) [reply red_blue_world/patches/gw.py]: Ohh I see. I actually had missed the fact that `coord` was a scalar and not a tuple. This makes sense now. No need to change it

I originally thought this was doing some bounds checking.

**Pull request merged**